python-lib/design_experiment/ab_dispatcher.py

```diff
@@ -47,60 +46,29 @@ def check_sample_size(self, experiment_population: np.array):
+    def split_into_groups(self, population_df: pd.DataFrame, leftovers_management: AttributionMethod) -> tuple:
         """Shuffle the population and splits it into groups
 
         :returns: Two arrays containing the contacts for each group
         :rtype: tuple
         """
         logger.info("Dispatching experiment population: shuffle")
-        np.random.permutation(experiment_population)
-        if attribution_method == AttributionMethod.LEFTOVER_TO_A:
-            B_group = experiment_population[:self.size_B]
-            A_group = experiment_population[self.size_B:]
-        elif attribution_method == AttributionMethod.LEFTOVER_TO_B:
-            A_group = experiment_population[:self.size_A]
-            B_group = experiment_population[self.size_A:]
-        elif attribution_method == AttributionMethod.LEFTOVER_BLANK:
-            B_group = experiment_population[:self.size_B]
-            A_group = experiment_population[self.size_B:self.size_A+self.size_B]
-        return A_group, B_group
-
-    def add_group_column(self, input_df: pd.DataFrame, A_group: np.array, B_group: np.array) -> pd.DataFrame:
-        """Add a column to the input dataframe to indicate the group for each user
-
-        :returns: The input dataframe with a "AB_group" column
-        :rtype: pd.DataFrame
-        """
-        input_df["AB_group"] = input_df[self.reference_column].apply(lambda x: retrieve_group(x, A_group, B_group))
-        return input_df
-
-    def group_to_df(self, group: np.array) -> pd.DataFrame:
-        """
-        Turns a group array into a one column dataframe
-
-        :param np.array group: array containing references to one experiment group (A or B)
-
-        :returns: A one column dataframe containing references to users of one grou
-        :rtype : pd.DataFrame
-        """
-        columns = [self.reference_column]
-        df = pd.DataFrame(data=group, index=range(group.shape[0]), columns=columns)
-        return df
-
-
-def retrieve_group(reference: str, A_group: np.array, B_group: np.array) -> Group:
-    """
-    Retrieve which group a user belongs to thanks to its reference
-
-    :param str reference: reference of a user (contact_id, email...)
-    :param np.array A_group : array containing the users of the A group
-    :param np.array B_group : array containing the users of the B group
-
-    :returns: The group where a user belongs
-    :rtype: Group
-    """
-    if reference in A_group:
-        return Group.A
-    elif reference in B_group:
-        return Group.B
+        shuffled_index = np.array(population_df.index)
+        if leftovers_management == AttributionMethod.LEFTOVER_TO_A:
+            population_df[Column.AB_GROUP.value] = Group.A.value
+            column_index = population_df.columns.get_loc(Column.AB_GROUP.value)
+            np.random.shuffle(shuffled_index)
+            population_df.iloc[shuffled_index[:self.size_B], column_index] = Group.B.value
```

Comment:
why not use `loc` directly

Reply:
done in https://github.com/dataiku/dss-plugin-ab-testing/pull/1/commits/30312e8d5a960f07c733d00a266a63ada524d92e